src/cmdstan/diagnose.cpp

```diff
@@ -0,0 +1,189 @@
+#include <algorithm>
+#include <iostream>
+#include <iomanip>
+#include <ios>
+#include <stan/mcmc/chains.hpp>
+#include <cmdstan/stansummary_helper.hpp>
+#include <fstream>
+
+void diagnose_usage() {
+  std::cout << "USAGE:  diagnose <filename 1> [<filename 2> ... <filename N>]"
+            << std::endl
+            << std::endl;
+}
+
+/**
+ * The Stan print function.
+ *
+ * @param argc Number of arguments
+ * @param argv Arguments
+ *
+ * @return 0 for success,
+ *         non-zero otherwise
+ */
+int main(int argc, const char* argv[]) {
+
+  if (argc == 1) {
+    diagnose_usage();
+    return 0;
+  }
+
+  // Parse any arguments specifying filenames
+  std::ifstream ifstream;
+  std::vector<std::string> filenames;
+
+  for (int i = 1; i < argc; i++) {
+    ifstream.open(argv[i]);
+
+    if (ifstream.good()) {
+      filenames.push_back(argv[i]);
+      ifstream.close();
+    } else {
+      std::cout << "File " << argv[i] << " not found" << std::endl;
+    }
+  }
+
+  if (!filenames.size()) {
+    std::cout << "No valid input files, exiting." << std::endl;
+    return 0;
+  }
+
+  // Parse specified files
+  ifstream.open(filenames[0].c_str());
+
+  stan::io::stan_csv stan_csv
+    = stan::io::stan_csv_reader::parse(ifstream, &std::cout);
+  stan::mcmc::chains<> chains(stan_csv);
+
+  ifstream.close();
+
+  for (std::vector<std::string>::size_type chain = 1;
+       chain < filenames.size(); chain++) {
+    ifstream.open(filenames[chain].c_str());
+    stan_csv = stan::io::stan_csv_reader::parse(ifstream, &std::cout);
+    chains.add(stan_csv);
+    ifstream.close();
+  }
+
+  int num_samples = chains.num_samples();
+
+  std::vector<std::string> bad_n_eff_names;
+  std::vector<std::string> bad_rhat_names;
+
+  for (int i = 0; i < chains.num_params(); i++) {
+    int max_limit = 10;
```

Comment:
It looks like this is related to max treedepth. Should this be controlled from the outside somehow?